File: Backend/app/schemas/workout_plans.py
# -*- coding: utf-8 -*-
from datetime import datetime

from pydantic import BaseModel


class Options(BaseModel):
    name: str
    image_url: str | None = None
    description: str
    muscle_group: str
    repetitions: str
    rest_time: int


class Plan(BaseModel):
    title: str
    exercises: list[Options]


class WorkoutPlanData(BaseModel):
    id: int
    plans: list[Plan] = []
    title: str
    description: str
    start_date: datetime
    end_date: datetime
    time_to_workout: str
    daily_training: int
    completed_days: int
    type: str


class WorkoutPlan(BaseModel):
    id: int
    title: str
    description: str
    plans: list[Plan] = []
    type: str


class ActualWorkoutPlanPrevious(BaseModel):
    id: int
    start_date: datetime
    end_date: datetime


class CreateWorkoutPlan(BaseModel):
    title: str
    description: str
    plans: list[Plan] = []
    is_public: bool | None = False
    months_valid: int
    days_per_week: int
    type: str
    time_to_workout: str | None = None
    user_id: int | None = None
    start_date: datetime | None = None


class AllFreeWorkoutPlanQuantity(BaseModel):
    quantity: int = 0


class ExpiringWorkoutPlans(BaseModel):
    id: int
    title: str
    user_id: int
    workout_id: int


class UpdateWorkoutPlan(BaseModel):
    title: str
    description: str
    plan: Plan
    is_public: bool | None = False
    months_valid: int
    last_update: datetime | None = datetime.now()


class LastFinishedWorkoutPlan(BaseModel):
    id: int
    title: str
    end_date: datetime


class PreviousWorkoutPlan(BaseModel):
    id: int
    title: str
    description: str
    professional_name: str | None = None


class WorkoutPlanCalendar(BaseModel):
    id: int
    start_date: datetime
    end_date: datetime
    days_per_week: int
    time_to_workout: str

File: Frontend/interfaces/workout_plans.ts
export interface WorkoutActualPrevious {
  id: number;
  start_date: Date;
  end_date: Date;
}

export interface Exercise {
  name: string;
  image_url?: string;
  muscle_group: string;
  rest_time: number;
  description: string;
  repetitions: string;
} 

export interface WorkoutPlan {
  title: string;
  exercises: Exercise[];
}

export interface CreateTraining {
  title: string;
  description: string;
  plans: WorkoutPlan[];
  is_public: boolean;
  days_per_week: number;
  type: string;
  months_valid: number;
  start_date?: string;
  user_id?: number;
  time_to_workout?: string;
}

export interface ViewWorkout {
  id: number
  title: string;
  description: string;
  professional_name?: string;
}

export interface ExpiringWorkout{
  id: number
  title: string;
  description: string;
  user_id: number;
  workout_id: number;
}

export interface WorkoutData {
    id: number;
    title: string;
    description: string;
    start_date: Date;
    end_date: Date;
    time_to_workout: string;
    daily_training: number;
    completed_days: number;
    type: string;
    plans: WorkoutPlan[];
}

export interface TrainingSchedule{
  id: number
  start_date: Date;
  end_date: Date;
  days_per_week: number;
  time_to_workout: string;
}